src/Features/VisualBasic/Portable/CodeRefactorings/SyncNamespace/VisualBasicSyncNamespaceService.vb

```diff
@@ -0,0 +1,79 @@
+﻿' Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+Imports System.Collections.Immutable
+Imports System.Composition
+Imports System.Threading
+Imports Microsoft.CodeAnalysis.CodeRefactorings.SyncNamespace
+Imports Microsoft.CodeAnalysis.Host.Mef
+Imports Microsoft.CodeAnalysis.VisualBasic.Syntax
+
+Namespace Microsoft.CodeAnalysis.VisualBasic.CodeRefactorings.SyncNamespace
+    <ExportLanguageService(GetType(ISyncNamespaceService), LanguageNames.VisualBasic), [Shared]>
+    Friend Class VisualBasicSyncNamespaceService
+        Inherits AbstractSyncNamespaceService(Of NamespaceStatementSyntax, CompilationUnitSyntax)
+
+        Public Overrides Function TryGetReplacementReferenceSyntax(reference As SyntaxNode, newNamespaceParts As ImmutableArray(Of String), ByRef old As SyntaxNode, ByRef [new] As SyntaxNode) As Boolean
+            Dim nameRef = TryCast(reference, SimpleNameSyntax)
+            If nameRef IsNot Nothing Then
+                Dim outerMostNode = GetQualifiedNameSyntax(nameRef)
+                old = outerMostNode
+
+                If outerMostNode Is nameRef Or newNamespaceParts.IsDefaultOrEmpty Then
+                    [new] = outerMostNode
+                Else
+                    If newNamespaceParts.Length = 1 And newNamespaceParts(0).Length = 0 Then
+                        [new] = nameRef
+                    Else
+                        Dim qualifiedNamespaceName = CreateNameSyntax(newNamespaceParts, newNamespaceParts.Length - 1)
+                        [new] = SyntaxFactory.QualifiedName(qualifiedNamespaceName, nameRef)
+                    End If
+                    [new] = [new].WithTriviaFrom(outerMostNode)
+                End If
+                Return True
+            Else
+                old = Nothing
+                [new] = Nothing
+                Return False
+            End If
+        End Function
+
+        Protected Overrides Function EscapeIdentifier(identifier As String) As String
+            Return identifier?.EscapeIdentifier()
+        End Function
+
+        ' This is only reachable when called from a VB refacoring provider, which is not implemented yet.
```

Comment:
??

This should be almost entirely language-agnostic.  hte only thing language specific is really doing a bit of the formatting around the updates.  i'm not sure why these aren't implemented.

Reply:
Imagine we trigger the refactoring to change a namespace in a C# document, which affect a reference to a type declared in it that is in a VB document. The part that fixes the reference in VB is language specific and can't be part of the C# version of the refactoring provider

Reply:
And yes, you are right that there's logic can be promoted to the abstract class. But I'd like to wait until the VB refactoring provider is implemented to clean all that up.

Reply:
> The part that fixes the reference in VB is language specific and can't be part of the C# version of the refactoring provider


My point though is that i don't see the need for *any* language specific code... it just makes things more brittle and less applicable to both languages.  If we write to our cross-language abstractions first, we pick up cross language support for effectively free ...

Reply:
Ah, you added more while i was writing that up.  

> But I'd like to wait until the VB refactoring provider is implemented to clean all that up.

Sure.  Can you ping me when the VB side is done.  I can take another pass.

In general though, i woudl say that it's really just good to write in a C#/VB agnostic way from teh start.  It usually just helps the code be clearer and removes a lot of specific code that isn't necessary.  Sometimes it's even possible to have the entire feature done without any language specific parts at all.  Then there's no need to even do 'the vb side' of things.  You just get both at the same time :)

Reply:
As mentioned earlier, no VB refactoring will be provided. So I will do this as part of #30896, when I have a better idea of the shape of the language service.